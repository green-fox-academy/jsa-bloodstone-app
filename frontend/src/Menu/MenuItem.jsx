import React from 'react';
import PropTypes from 'prop-types';
import {
  Text, Image, View, TouchableHighlight, Dimensions, StyleSheet,
} from 'react-native';
import Colors from '../common/colors';

const vw = Math.round(Dimensions.get('window').width / 600);
const FONT_SIZE = 12 * vw;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    borderRadius: 6,
    alignItems: 'center',
    justifyContent: 'center',
  },
  container2: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    width: '100%',
    paddingVertical: 8,
  },
  textStyle: {
    marginTop: 5,
    fontSize: FONT_SIZE,
    fontWeight: 'bold',
  },
});

function MenuItem({
  src, name, onPress, activeComponent,
}) {
  const containerColorStyle = {
    backgroundColor: activeComponent === name
      ? Colors.lightenGrey
      : Colors.transparent,
  };
  const textColorStyle = {
    color: activeComponent === name
      ? '#333'
      : '#bdbdbd',
  };
  return (
    <TouchableHighlight
      style={[styles.container, containerColorStyle]}
      onPress={() => onPress(name)}
      underlayColor={Colors.white20Color}
    >
      <View style={styles.container2}>
        <Image style={{ flex: 1 }} source={src} resizeMode="contain" />
        <Text numberOfLines={1} style={[styles.textStyle, textColorStyle]}>{name}</Text>
      </View>
    </TouchableHighlight>
  );
}

MenuItem.propTypes = {
  src: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  onPress: PropTypes.func,
  activeComponent: PropTypes.string,
};

MenuItem.defaultProps = {
  onPress: null,
  activeComponent: '',
};

export default MenuItem;
