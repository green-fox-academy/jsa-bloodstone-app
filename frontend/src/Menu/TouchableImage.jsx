import React from 'react';
import PropTypes from 'prop-types';
import {
  Text, Image, View,
  TouchableHighlight,
  Dimensions, StyleSheet,
} from 'react-native';

const vw = Math.round(Dimensions.get('window').width / 600);
const FONT_SIZE = 12 * vw;

const styles = StyleSheet.create({
  menuItemContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    width: '100%',
    paddingVertical: 12,
  },
  menuText: {
    marginTop: 5,
    fontSize: FONT_SIZE,
    fontWeight: 'bold',
    color: '#55565a',
  },
});

function TouchableImage({
  src, name, onPress, activeComponent,
}) {
  return (
    <TouchableHighlight
      underlayColor={Colors.lightenTeal}
      onPress={() => onPress(name)}
      style={{
        flex: 1,
        borderRadius: 6,
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: activeComponent === name
          ? Colors.lightenTeal
          : 'transparent',
      }}
    >
      <View style={styles.menuItemContainer}>
        <Image
          source={src}
          resizeMode="contain"
          style={{ flex: 1 }}
        />
        <Text numberOfLines={1} style={styles.menuText}>{name}</Text>
      </View>
    </TouchableHighlight>
  );
}

TouchableImage.propTypes = {
  src: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  onPress: PropTypes.func,
  activeComponent: PropTypes.string,
};

TouchableImage.defaultProps = {
  onPress: null,
  activeComponent: '',
};

export default TouchableImage;
