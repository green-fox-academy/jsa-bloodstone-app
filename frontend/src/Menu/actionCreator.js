export const CHANGE_DISPLAYED_COMPONENT = 'changeDisplayedComponent';

export function changeDisplayedComponent(pageSelected) {
  return {
    type: CHANGE_DISPLAYED_COMPONENT,
    payload: pageSelected,
  };
}
