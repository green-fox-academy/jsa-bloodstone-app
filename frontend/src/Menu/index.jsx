import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { StyleSheet } from 'react-native';
import { CardView } from '../common/components';
import { changeDisplayedComponent } from './actionCreator';
import MenuItem from './MenuItem';
import buildingMenuIcon from '../../assets/menu/Buildings.png';
import troopsMenuIcon from '../../assets/menu/Troops.png';
import battleMenuIcon from '../../assets/menu/Battle.png';
import leaderBoardMenuIcon from '../../assets/menu/Leaderboard.png';
import Colors from '../common/colors';

const ICON_LIST = [
  { name: 'Buildings', url: buildingMenuIcon },
  { name: 'Troops', url: troopsMenuIcon },
  { name: 'Battle', url: battleMenuIcon },
  { name: 'Leaderboard', url: leaderBoardMenuIcon },
];

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    height: 120,
    paddingVertical: 10,
    backgroundColor: Colors.whiteColor,
  },
});

function Menu() {
  const activeComponent = useSelector((state) => state.menu.currentlyDisplayComponent);
  const dispatch = useDispatch();

  function handlePress(name) {
    dispatch(changeDisplayedComponent(name));
  }

  return (
    <CardView style={styles.container}>
      {ICON_LIST.map((element) => (
        <MenuItem
          key={element.url}
          src={element.url}
          name={element.name}
          onPress={handlePress}
          activeComponent={activeComponent}
        />
      ))}
    </CardView>
  );
}

export default Menu;
