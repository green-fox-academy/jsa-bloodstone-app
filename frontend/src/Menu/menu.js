import { CHANGE_DIAPLAYED_COMPONENT } from './actionCreator';

const initialState = {
  currentlyDisplayComponent: 'Buildings',
};

export default function menu(state = initialState, action) {
  switch (action.type) {
    case CHANGE_DIAPLAYED_COMPONENT:
      return { ...state, currentlyDisplayComponent: action.payload };
    default:
      return { ...state, currentlyDisplayComponent: 'Buildings' };
  }
}
