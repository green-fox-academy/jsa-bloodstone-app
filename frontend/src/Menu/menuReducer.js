import { CHANGE_DISPLAYED_COMPONENT } from './actionCreator';

const initialState = {
  currentlyDisplayComponent: 'Buildings',
};

export default function menu(state = initialState, action) {
  switch (action.type) {
    case CHANGE_DISPLAYED_COMPONENT:
      return { ...state, currentlyDisplayComponent: action.payload };
    default:
      return state;
  }
}
