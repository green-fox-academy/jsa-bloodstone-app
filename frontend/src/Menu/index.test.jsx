import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import Menu from '.';

jest.mock('react-redux');
jest.mock('./actionCreator');

describe('<Menu />', () => {
  it('Should render menu and match the snapshot', () => {
    const renderer = new ShallowRenderer();
    const component = renderer.render(<Menu />);
    expect(component).toMatchSnapshot();
  });
});
