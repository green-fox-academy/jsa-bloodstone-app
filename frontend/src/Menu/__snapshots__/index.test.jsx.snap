// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> Should render menu and match the snapshot 1`] = `
<CardView
  style={
    Object {
      "height": 150,
      "marginBottom": 0,
      "padding": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "rgba(0,0,0,0.08)",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginHorizontal": 10,
      }
    }
  >
    <View
      style={
        Object {
          "paddingVertical": 7,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        My Kingdom
      </Text>
    </View>
    <Resources />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "padding": 10,
      }
    }
  >
    <MenuItem
      activeComponent=""
      name="Buildings"
      onPress={[Function]}
      src={1}
    />
    <MenuItem
      activeComponent=""
      name="Troops"
      onPress={[Function]}
      src={1}
    />
    <MenuItem
      activeComponent=""
      name="Battle"
      onPress={[Function]}
      src={1}
    />
    <MenuItem
      activeComponent=""
      name="Leaderboard"
      onPress={[Function]}
      src={1}
    />
  </View>
</CardView>
`;
